topic-operator/src/main/java/io/strimzi/operator/topic/KafkaStreamsTopicStoreService.java

```diff
@@ -0,0 +1,192 @@
+
+        this.store = new KafkaStreamsTopicStore(sc.getStore(), storeTopic, producer, sc.getService());
+        return this.store;
+    }
+
+    private CompletableFuture<AsyncBiFunctionService.WithSerdes<String, String, Integer>> createKafkaStreams(Config config, Properties kafkaProperties, String storeTopic, String storeName) {
+        long timeoutMillis = config.get(Config.STALE_RESULT_TIMEOUT_MS);
+        ForeachActionDispatcher<String, Integer> dispatcher = new ForeachActionDispatcher<>();
+        WaitForResultService serviceImpl = new WaitForResultService(timeoutMillis, dispatcher);
+        closeables.add(serviceImpl);
+
+        AtomicBoolean done = new AtomicBoolean(false); // no need for dup complete
+        CompletableFuture<AsyncBiFunctionService.WithSerdes<String, String, Integer>> cf = new CompletableFuture<>();
+        KafkaStreams.StateListener listener = (newState, oldState) -> {
+            if (newState == KafkaStreams.State.RUNNING && !done.getAndSet(true)) {
+                cf.completeAsync(() -> serviceImpl); // complete in a different thread
+            }
+            if (newState == KafkaStreams.State.ERROR) {
+                cf.completeExceptionally(new IllegalStateException("KafkaStreams error"));
+            }
+        };
+
+        Topology topology = new TopicStoreTopologyProvider(storeTopic, storeName, kafkaProperties, dispatcher).get();
+        streams = new KafkaStreams(topology, kafkaProperties);
+        streams.setStateListener(listener);
+        streams.setGlobalStateRestoreListener(new LoggingStateRestoreListener());
+        closeables.add(streams);
+        streams.start();
+
+        return cf;
+    }
+
+    private CompletionStage<Void> createNewStoreTopic(String storeTopic, Admin admin, Context c) {
+        int rf = Math.min(3, c.clusterSize);
+        int minISR = Math.max(rf - 1, 1);
+        NewTopic newTopic = new NewTopic(storeTopic, 1, (short) rf)
+            .configs(Collections.singletonMap(TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG, String.valueOf(minISR)));
+        return toCS(admin.createTopics(Collections.singleton(newTopic)).all());
+    }
+
+    private CompletionStage<Void> validateExistingStoreTopic(String storeTopic, Admin admin, Context c) {
+        ConfigResource storeTopicConfigResource = new ConfigResource(ConfigResource.Type.TOPIC, storeTopic);
+        return toCS(admin.describeTopics(Collections.singleton(storeTopic)).values().get(storeTopic))
+            .thenApply(td -> c.setRf(td.partitions().stream().map(tp -> tp.replicas().size()).min(Integer::compare).orElseThrow()))
+            .thenCompose(c2 -> toCS(admin.describeConfigs(Collections.singleton(storeTopicConfigResource)).values().get(storeTopicConfigResource))
+                    .thenApply(cr -> c2.setMinISR(parseInt(cr.get(TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG).value()))))
+            .thenApply(c3 -> {
+                if (c3.rf != Math.min(3, c3.clusterSize) || c3.minISR != c3.rf - 1) {
```

Comment:
So I was thinking about this scenario ...
Initially, the cluster is made by 3 brokers so we are going to create the store topic with rf = 3 and minISR=2 (as per `createNewStoreTopic` method).
Then imagine that we scale up to 9 as far as I understood from this code, the store topic will remain with same configuration. I am not sure that it will be for production use anymore, with a cluster of 9 but a topic with rf = 3 and minISR=2. Shouldn't we increase the replication factor accordingly? @tombentley thoughts?

Reply:
This is what @tombentley suggested ... dunno if you really need (rf == cluster size) with big clusters ...

Reply:
Maybe not exactly cluster size but something more close to it

Reply:
I'll let @tombentley work out the rf value algorithm